lib/ansible/modules/cloud/cloudstack/cs_physical_network.py

```diff
@@ -0,0 +1,470 @@
+        return vrouter
+
+    def set_loadbalancer_element_state(self, enabled, nsp_name='InternalLbVm'):
+        loadbalancer = self.get_loadbalancer_element(nsp_name=nsp_name)
+        if loadbalancer['enabled'] == enabled:
+            return loadbalancer
+
+        args = {
+            'id': loadbalancer['id'],
+            'enabled': enabled
+        }
+        res = self.query_api('configureInternalLoadBalancerElement', **args)
+
+        poll_async = self.module.params.get('poll_async')
+        if poll_async:
+            loadbalancer = self.poll_job(res, 'internalloadbalancerelement')
+
+        self.result['changed'] = True
+        return loadbalancer
+
+    def present_network(self):
+        network = self.get_physical_network()
+        if network:
+            network = self._update_network()
+        else:
+            network = self._create_network()
+        return network
+
+    def _create_network(self):
+        self.result['changed'] = True
+        args = dict(zoneid=self.get_zone(key='id'))
+        args.update(self._get_common_args())
+        if self.get_domain(key='id'):
+            args['domainid'] = self.get_domain(key='id')
+
+        if not self.module.check_mode:
+            resource = self.query_api('createPhysicalNetwork', **args)
+
+            poll_async = self.module.params.get('poll_async')
+            if poll_async:
+                self.network = self.poll_job(resource, 'physicalnetwork')
+
+        return self.network
+
+    def _update_network(self):
+        network = self.get_physical_network()
+
+        args = dict(id=network['id'])
+        args.update(self._get_common_args())
+
+        if self.has_changed(args, network):
+            self.result['changed'] = True
+
+            if not self.module.check_mode:
+                resource = self.query_api('updatePhysicalNetwork', **args)
+
+                poll_async = self.module.params.get('poll_async')
+                if poll_async:
+                    self.physical_network = self.poll_job(resource, 'physicalnetwork')
+        return self.physical_network
+
+    def absent_network(self):
+        physical_network = self.get_physical_network()
+        if physical_network:
+            self.result['changed'] = True
+            args = {
+                'id': physical_network['id'],
+            }
+            if not self.module.check_mode:
+                resource = self.query_api('deletePhysicalNetwork', **args)
+                poll_async = self.module.params.get('poll_async')
+                if poll_async:
+                    self.poll_job(resource, 'success')
+
+        return physical_network
+
+
+def main():
+    argument_spec = cs_argument_spec()
+    argument_spec.update(dict(
+        physical_network=dict(required=True, aliases=['name']),
+        zone=dict(default=None),
+        domain=dict(default=None),
+        vlan=dict(default=None),
+        nsps_disabled=dict(default=None, type='list'),
```

Comment:
Wouldn't it be possible to add a choice list of possible Network Service Providers?

Reply:
I'm quite new in the Cloudstack world so I really don't know how dynamic that list actually is between deployments. NSP validation is done by a separate call to the API. I'm not sure if it's a static list or a dynamic one depending on the installed/enabled plugins.

Reply:
OK, not a blocker anyway.